src/cargo/util/semver_ext.rs

```diff
@@ -116,7 +116,7 @@ impl OptVersionReq {
     /// The logic here is temporary, we'll have to consider more boundary conditions later,
     /// and we're not sure if this part of the functionality should be implemented in semver or cargo.
     pub fn matches_prerelease(&self, version: &Version) -> bool {
-        if version.is_prerelease() {
+        if version.is_prerelease() && matches!(self, OptVersionReq::Req(_)) {
```

Comment:
nit: to make the semantics more clear, we should have `matches!(self, OptVersionReq::Req(_))` on the outside (left side) as that is guarding our overriding of the   `matches` logic to get the pre-release behavior.